tests/unit/query_model/test_nodes.py

```diff
@@ -453,13 +453,55 @@ class SomeInternalOperation(Series):
     assert SomeInternalOperation(value=mixed_set)
 
 
-def test_comparisons_between_value_nodes_are_strict():
-    assert Value(10) == Value(10)
-    assert Value(10) != Value(10.0)
-    assert Value(1) != Value(True)
+@pytest.mark.parametrize(
+    "lhs,cmp,rhs",
+    [
+        (10, "==", 10),
+        (10, "!=", 11),
+        (10, "!=", 10.0),
+        (True, "==", True),
+        (True, "!=", False),
+        (1, "!=", True),
+        (1.0, "!=", True),
+        (0, "!=", False),
+        (0.0, "!=", False),
+        # Test that we're strict about booleans embedded in frozensets
+        (frozenset({True}), "==", frozenset({True})),
+        (frozenset({True}), "!=", frozenset({False})),
+        (frozenset({1}), "!=", frozenset({True})),
+        (frozenset({1.0}), "!=", frozenset({True})),
+        (frozenset({0}), "!=", frozenset({False})),
+        (frozenset({0.0}), "!=", frozenset({False})),
+        (frozenset({True, False}), "==", frozenset({False, True})),
+        # Test the fast path for identical frozensets
+        (example := frozenset({1, 2, 3}), "==", example),
+        # Test the fast path for different sized frozensets
+        (frozenset({1, 2, 3}), "!=", frozenset({1, 2})),
+        # Test the normal path for frozenset equality
+        (frozenset({1, 2, 3}), "==", frozenset({3, 2, 1})),
+        (frozenset({1, 2, 3}), "!=", frozenset({4, 5, 6})),
+    ],
+)
+def test_comparisons_between_value_nodes_are_strict(lhs, cmp, rhs):
+    if cmp == "==":
```

Comment:
:heart:

Reply:
Could we not use `operator` here? We do when testing the query language. For example:

https://github.com/opensafely-core/ehrql/blob/ea07908488e07c7f84b14a120aa428eab3da8d98/tests/unit/test_query_language.py#L387-L403

Reply:
Yeah we could. Those QL tests are part of a whole series of similar tests which use a whole bunch of different operators so that definitely seems like the right approach there. For just `==` and `!=` this way felt a bit more obvious and simple to me, but I wouldn't object to them being done either way really.